File: Postfix-Lua/Postfix.lua
local function first(s)
    return s:sub(1, 1)
end
local function rest(s)
    return s:sub(2)
end

local zero = string.byte'0'
local nine = string.byte'9'

local function numberEnd(s)
    for i = 1, #s do
        local char = s:byte(i)
        if not (zero <= char and char <= nine) then
            return i
        end
    end
    return nil
end

local function number(tokens, s)
    local position = numberEnd(s)
    local theNumber, tail = s, ""
    if position then
        theNumber, tail = s:sub(1, position-1), s:sub(position)
    end
    tokens[#tokens+1] = theNumber
    return tail
end

local function factor(tokens, s)
    if '(' == first(s) then
        local tail = expression(tokens, rest(s))
        assert(')' == first(tail))
        return rest(tail)
    end
    return number(tokens, s)
end

local function binaryOperation(tokens, s, operator, subproduction)
    local tail = subproduction(tokens, s)
    while operator == first(tail) do
        tail = subproduction(tokens, rest(tail))
        tokens[#tokens+1] = operator
    end
    return tail
end

local function term(tokens, s)
    return binaryOperation(tokens, s, '*', factor)
end

function expression(tokens, s)
    return binaryOperation(tokens, s, '+', term)
end

function postfix(s)
    local tokens = {}
    expression(tokens, s)
    return table.concat(tokens, ' ')
end

File: Postfix-Haskell/Postfix.hs
module Postfix where
import Data.List
import Char

type Token = String
type Rest = String

data TokensAndRest = TokensAndRest {
      tokens :: [Token],
      rest :: Rest
    } deriving (Eq, Show)

type TokenAccumulator = (TokensAndRest -> TokensAndRest)

postfix s = tokens $ expression $ TokensAndRest [] s

-- expression ::= <term> ["+" <term>]*

expression :: TokenAccumulator
expression x = loopOnPlus $ term x

loopOnPlus :: TokenAccumulator
loopOnPlus (TokensAndRest tokens ('+':cs)) = loopOnPlus $ TokensAndRest tokens'' cs'
    where TokensAndRest tokens' cs' = term $ TokensAndRest tokens cs
          tokens'' = tokens' ++ ["+"]
loopOnPlus x = x

-- term ::= <factor> ["*" <factor>]*

term :: TokenAccumulator
term x = loopOnStar $ factor x

loopOnStar :: TokenAccumulator
loopOnStar (TokensAndRest tokens ('*':cs)) = loopOnStar $ TokensAndRest tokens'' cs'
    where TokensAndRest tokens' cs' = factor $ TokensAndRest tokens cs
          tokens'' = tokens' ++ ["*"]
loopOnStar x = x

-- binaryOperation :: Char -> TokenAccumulator -> TokenAccumulator
-- binaryOperation operator subexpression = f
--     where f :: TokenAccumulator
--           f (TokensAndRest tokens (operator:cs)) = f $ TokensAndRest tokens'' cs'
--               where TokensAndRest tokens' cs' = subexpression $ TokensAndRest tokens cs
--                     tokens'' = tokens' ++ [[operator]]
--           f x = x

-- factor ::= "(" <expression> ")"
--            | <number>

factor :: TokenAccumulator
factor (TokensAndRest tokens ('(':cs)) = TokensAndRest tokens' cs''
    where TokensAndRest tokens' cs' = expression $ TokensAndRest tokens cs
          cs'' = tail cs'
factor x = number x

-- number ::= ["0".."9"]+

number :: TokenAccumulator
number (TokensAndRest tokens s) = TokensAndRest (tokens ++ [n]) rest
    where (n, rest) = span isDigit s
